allennlp/modules/stacked_elmo_lstm.py

```diff
@@ -83,16 +89,78 @@ def __init__(self,
         self.backward_layers = backward_layers
 
     def forward(self,  # pylint: disable=arguments-differ
-                inputs: PackedSequence,
-                initial_state: Optional[Tuple[torch.Tensor, torch.Tensor]] = None) -> Tuple[torch.Tensor,
-                                                                                            Tuple[torch.Tensor,
-                                                                                                  torch.Tensor]]:
+                inputs: torch.Tensor,
+                mask: torch.LongTensor) -> torch.Tensor:
+        """
+        Parameters
+        ----------
+        inputs : ``torch.Tensor``, required.
+            A Tensor of shape ``(batch_size, sequence_length, hidden_size)``.
+        mask : ``torch.LongTensor``, required.
+            A binary mask of shape ``(batch_size, sequence_length)`` representing the
+            non-padded elements in each sequence in the batch.
+
+        Returns
+        -------
+        A ``torch.Tensor`` of shape (num_layers, batch_size, sequence_length, hidden_size),
+        where the num_layers dimension represents the LSTM output from that layer.
+        """
+        batch_size, total_sequence_length = mask.size()
+        stacked_sequence_output, final_states, restoration_indices, num_valid = \
+            self.sort_and_run_forward(self.lstm_forward, inputs, mask)
+
+        # stacked_sequence_output is shape (num_layers, batch_size, timesteps, encoder_dim)
+        num_layers = stacked_sequence_output.size(0)
+        per_layer_sequence_outputs = [layer.squeeze(0) for layer in
```

Comment:
Is this more efficient than `[layer[i] for i in num_layers]`?  If it is, I need to update my decoding code...

Reply:
But why are you doing this at all?  Why can't you just operate on the 4D tensor?  Is it because of the `index_select` at the end?  If that's the case, you can still do all of the padding stuff on the 4D tensor, which will result in fewer tensor operations, and just do a chunk, index select, and stack at the end (or maybe there's one of the fancy `batched_index_select` methods from the coref PR that would do this without needing the chunk and stack).